usr.sbin/smtpd/parse.y

```diff
@@ -3465,10 +3466,12 @@ host_dns(struct listen_opts *lo)
 		return (-1);
 	}
 
-	for (res = res0; res; res = res->ai_next) {
+	for (res = res0; res; prevaddr = res->ai_addr, res = res->ai_next) {
 		if (res->ai_family != AF_INET &&
 		    res->ai_family != AF_INET6)
 			continue;
+		if (prevaddr != NULL && memcmp(prevaddr, res->ai_addr, sizeof(struct sockaddr)) == 0)
```

Comment:
I believe that rather than `sizeof(struct sockaddr)` we should use `res->ai_addrlen`, i.e.

```
if (prevaddr != NULL &&
    prevaddr->ai_addrlen == res->ai_addrlen &&
    memcmp(prevaddr->ai_addr, res->ai_addr, res->ai_addrlen) == 0)
	continue;
```

Also, while here, I guess to stay on the safer side we could loop over `conf->sc_listeners` and check for duplicates there.  (imagine the case where we first add `127.0.0.1`, then `::1`, and finally `127.0.0.1` again.)